[GeoMechanicsApplication] Moved logic of `ApplyScalarConstraintsTableProcess` to C++

avdg81 opened the pull request:
**📝 Description**
This class used to be defined in a Python module only. Now that we also need it for the custom settlement workflow, we had to port it from Python to C++. Note that the Python interface now uses the new C++ class.

Also added a creator for the `ApplyScalarConstraintsTableProcess` to the process factory used by the custom workflow. The factory should now be able to create all kinds of processes that are relevant for settlement analysis.

indigocoral committed:
Moved logic of `ApplyScalarConstraintsTableProcess` to C++

This class used to be defined in a Python module only. Now that we also
need it for the custom settlement workflow, we had to port it from
Python to C++. Note that the Python interface now uses the new C++
class.

Also added a creator for the `ApplyScalarConstraintsTableProcess` to the
process factory used by the custom workflow. The factory should now be
able to create all kinds of processes that are relevant for settlement
analysis.

rfaasse commented on applications/GeoMechanicsApplication/custom_processes/apply_scalar_constraints_table_process.cpp:
```diff
@@ -0,0 +1,227 @@
+{
+    NamesOfSettingsToCopy.emplace_back("value");
+
+    if (rProcessSettings["table"].GetInt() == 0) {
```
`rProcessSettings["table"].GetInt() == 0` is used a lot. We could consider making it a function with a descriptive name (then it's also clearer what this check means, because to me that's not clear and if the hardcoded "table" would ever change, we only have to change it in one place)

avdg81 replied:
Done.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_processes/apply_scalar_constraints_table_process.cpp:
```diff
@@ -0,0 +1,227 @@
+    }
+}
+
+void ApplyScalarConstraintsTableProcess::MakeProcessForPhreaticSurface(const Parameters&        rProcessSettings,
```
This one belongs together with or close to the line and hydrostatic for functional reasons. The interpolate line process is a deviating functionality.

avdg81 replied:
Done together :-)

WPK4FEM commented on applications/GeoMechanicsApplication/custom_processes/apply_scalar_constraints_table_process.cpp:
```diff
@@ -0,0 +1,227 @@
+    }
+}
+
+void ApplyScalarConstraintsTableProcess::MakeProcessForUniformFluidPressure(const Parameters&        rProcessSettings,
```
this to me looks exactly the same as the implementations of MakeProcessForNonFluidPressureType. Why have 2.

avdg81 replied:
Done.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_processes/apply_scalar_constraints_table_process.cpp:
```diff
@@ -0,0 +1,227 @@
+    AppendParameterNameIfExists("pressure_tension_cut_off", rProcessSettings, NamesOfSettingsToCopy);
+    AppendParameterNameIfExists("is_seepage", rProcessSettings, NamesOfSettingsToCopy);
+
+    if (rProcessSettings["table"].GetInt() == 0) {
```
The test for a table on hydrostatic, phreatic line and phreatic surface looks like any of an array that has length 1, 2 or 3 is unequal to zero. The test of using a table is more natural to me than not using a table. Can we put the test if table functionality should be used in a function?

avdg81 replied:
Done.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_processes/apply_scalar_constraints_table_process.cpp:
```diff
@@ -0,0 +1,227 @@
+    }
+}
+
+void ApplyScalarConstraintsTableProcess::MakeProcessForNonFluidPressureType(const Parameters&        rProcessSettings,
```
This is the general thing, the fluid pressures the specialized thing. I would prefer if the general thing has a general name and does not refer to not being the special thing.

avdg81 replied:
We have renamed the member function, so done.

rfaasse commented on applications/GeoMechanicsApplication/custom_processes/apply_scalar_constraints_table_process.cpp:
```diff
@@ -0,0 +1,227 @@
+    AppendParameterNameIfExists("is_fixed", rProcessSettings, names_of_settings_to_copy);
+
+    if (rProcessSettings.Has("fluid_pressure_type")) {
+        MakeProcessForFluidPressureType(rProcessSettings, std::move(names_of_settings_to_copy));
```
I'm not sure if I agree with the necessity of using std::move for a vector with only two strings. I think passing it as a ref would be fine as well and a bit less confusing, but I leave it up to you, both ways are acceptable to me.

avdg81 replied:
I am stubborn and insist on using `std::move` here ;-). My primary motivation is to show that I want to change the scope of the list of setting names.

WPK4FEM reviewed (commented):
Thank you, this is more reusable than the python interface.

rfaasse commented on applications/GeoMechanicsApplication/custom_processes/apply_scalar_constraints_table_process.h:
```diff
@@ -0,0 +1,63 @@
+//
+#pragma once
+
+#include "processes/process.h"
```
I think here Process can also be forward declared (like ModelPart and Parameters)

avdg81 replied:
We tried, but this class derives from `Process`, so no ;-)

rfaasse reviewed (dismissed):
Clear and clean code! I have some minor suggestions, nothing blocking. 

On the point of UTs: looking at the class it's not soo trivial to unit-test and since it's already tested by the python tests, I don't think it is worth the effort at this point in time

avdg81 committed:
Started to process Wijtze Pieter's review comments

- Moved a member function to a more logical place.
- Removed a duplicated member function and renamed the remaining one.
- Added a helper function to detect whether a table has been attached to
  a process.
- Since the logic was inverted, some statements had to be rearranged.

rfaasse reviewed (approved):
Looks good, thanks for incorporating the review comments!
